crates/pairwise_nostd_internal/src/parallel.rs

```diff
@@ -0,0 +1,91 @@
+//! Our parallelism abstractions use the concepts of thread teams & reductions
+
+use core::num::NonZeroU32;
+use ndarray::{ArrayView1, ArrayView2, ArrayViewMut1, ArrayViewMut2};
+
+pub struct TeamRank(u32);
+
+impl TeamRank {
+    pub fn new(rank: u32) -> Self {
+        TeamRank(rank)
+    }
+
+    pub fn get(&self) -> u32 {
+        self.0
+    }
+}
+
+/// the goal here is to use a parallelism strategy with thread-teams. This is
+/// hardly a novel concept. The actual implementation will be tailored to the
+/// "backend" that we are using (e.g. serial, multi-threading, GPU)
+///
+/// To make things work, we borrow the terminology used by Kokkos:
+/// we have a league of `league_size` thread-teams. Each team is composed
+/// `thread_size` threads.
+pub trait TeamProps {
+    // for now, we explicitly choose not to provide a method to directly access
+    // team_rank. The current approach for accessing team_rank (i.e. only as
+    // part of `team_reduce`) generally encourages a design where the serial
+    // implementation can support arbitrary team sizes (and is capable of
+    // returning the exact same result as parallel implementations). If we
+    // make team_rank accessible, implementations can generally only support
+    // a team_size of 1.
+
+    fn team_size(&self) -> u32;
+    fn league_rank(&self) -> u32;
+    fn league_size(&self) -> u32;
+
+    /// Triggers a barrier for all threads in the thread team, then has each
+    /// team-member invoke `prep_buf_fn` to prepare separate input buffers
+    /// used in for a reduction. Afterwards, `reduce_buf_fn` is used to perform
+    /// reductions with all of the buffers
+    ///
+    /// `prep_buf_fn` and `reduce_buf_fn` generally have the signatures:
+    /// ```text
+    /// prep_buf_fn(reduce_buf: &mut ArrayViewMut1<f64>, team_rank: TeamRank)
+    /// reduce_buf_fn(reduce_buf: &mut ArrayViewMut1<f64>, other_buf: ArrayView1<f64>)
+    /// ```
```

Comment:
Is this redundant with the type signature?

Reply:
The type signature doesn't make it clear what the arguments actually mean. This is mostly an issue for `reduce_buf_fn`.